### DISCUSSION TITLE
authorization

### DISCUSSION THREAD

**Discussion by ltorresu82** [honojs/hono#2452] [General]
Is it possible to protect routes using for example casbin?

**Comment by linhub15** [comment]
Yes you can protect routes by using middleware. Basic, Bearer, JWT middleware is built in but for casbin you may have to create your own.

https://hono.dev/middleware/builtin/basic-auth
```ts
import { Hono } from 'hono'
import { basicAuth } from 'hono/basic-auth'
const app = new Hono()
app.get(
  '/secure',
  basicAuth({ 'admin', 'axkdu2w23' }),
  (c) => {
    return c.text('You are authorized')
  }),
)
```

**Reply by ltorresu82** [reply]
Yes, that is in the documentation. But what I mean is that you can discriminate access to a route by role and/or permissions. ¿?

**Reply by NicoPlyley** [reply]
I don't know if casbin would be compatible, I tried looking it up but didn't see any clear signs. But for making access based off a role is very easy to setup. I have an auth middleware and I pass it the roles I want to allow. Eg. `auth('admin', 'editor')` and it checks against the user and only lets those roles access the route. Otherwise throws a 401

**Reply by linhub15** [reply]
From what I see in casbin, there is no provided middleware for hono, but here are some of their other middleware you can use as an example.
- [express-authz](https://github.com/node-casbin/express-authz/blob/master/src/authz.ts)
- [koa-authz](https://github.com/node-casbin/koa-authz/blob/master/authz.js)

Then you can use the custom middleware on any route.
```ts
app.get(
  '/secure',
  casbinAuthz(<roles or permissions>),
   (c) => {
    return c.text('You are authorized')
  }),
)
```
